enhancements/platforms/ovirt-rhv-ipi.md

```diff
@@ -0,0 +1,271 @@
+last-updated: 2019-10-07
+status: implementable
+---
+
+# ovirt-rhv-platform-provider
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+## Summary
+
+This document describes how `oVirt` becomes a platform provider for Openshift. \
+`oVirt` is a virtualization platform and is similar to the `baremetal` platform \
+ provider it is lacking DNS and Load-Balancing services, but it has the advantage \
+if software-defining your datacenter, utilizing existing hardware and making \ 
+that pain free and fast.
+Like `baremetal` platform it uses the [OpenShift-Metal³ kni-installer](https://github.com/openshift-metal3/kni-installer) method \
+- essentially providing an internal cluster-level DNS service using mDNS and
+ coreDNS, and load-balancing using keepalived.
+
+Components involved:
+  - github.com/openshift/api
+  - github.com/openshift/installer
+  - github.com/openshift/machine-config-operator
+  - github.com/openshift/cluster-image-registry-operator (@adambkapln?)
+  - github.com/ovirt/cluster-api-provider-ovirt
+  - github.com/ovirt/terraform-provider-ovirt
+
+## Motivation
+
+- It has been defined as a key initiative for 4.3
+- The feedback RHV team got is that there is a lot of demand for this 
+  kind of installation
+
+### Goals
+
+- provide a way to install Openshift on oVirt infrastructure using 
+  the installer - an IPI installation.
+- implementing a cluster-api provider to provide scaling and managing the cluster
+  nodes (used by IPI, and useful for UPI, and also node management/fencing)
+
+### Non-Goals
+- UPI implementation will be provided separately.
+
+## Proposal
+
+This provider enables the Openshift Installer to provision VM resources in an \
+oVirt data center, that will be used as worker and masters of the clusters. It \
+will also create the bootstrap machine, and the configuration needed to get \
+the initial cluster running by supplying DNS a service and load balancing, all \
+using static pods.
+
+This work is related to the Bare-Metal provider because oVirt does not supply \
+DNS and LB services but is a platform provider. See the [baremetal ipi networking doc][baremetal-ipi-networking]
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. Survey
+
+The installation starts and right after the user supplies his public ssh key,\
+and then choose `ovirt` the installation will ask for all the relevant details\
+of the installation: **url**, **user** and **password** of ovirt api. The installer \
+will validate it can communicate with the api, otherwise it will fail to proceed.\
+Next the installer will ask to choose the ovirt **cluster**, where the VMs will be\
+created, and next would be a *template* from that cluster for those VMs.\
```

Comment:
Can you link or provide details around what is a template?

Reply:
done

> I would like to see these things detailed:
> 
>     1. How does the boot-images for VMs work?
> 
Will adding this document help? [coreos ignition support][https://ovirt.org/develop/release-management/features/virt/coreos-ignition-support.html]

>     2. How the ingress-controllers are setup, internally for cluster and how it would be exposed for users.

> 
>     3. What is the storage setup.
> 

As far as I understand this is day-two operation needed for the registry?


>     4. Does is turn on any cloud-provider code in k8s? If yes, how are credentials handled?
> 
no, cloud provider is not needed.

> 
> Also the docs references `Baremetal networking` and I couldn't find link for that..

It is at the bottom

Reply:
> https://ovirt.org/develop/release-management/features/virt/coreos-ignition-support.html

404s

> 1. How does the boot-images for VMs work?

Where will you get the boot-images from? For bootstrap, control-plane and compute.
Is there anything special when providing large ignition files like bootstrap.ign

Reply:
> > https://ovirt.org/develop/release-management/features/virt/coreos-ignition-support.html
> 
> 404s

It works

> 
> > 1. How does the boot-images for VMs work?
> 
> Where will you get the boot-images from? For bootstrap, control-plane and compute.

For this stage the users will upload the RHCOS qcow image as a template. Later in the survey they will see it and choose it. I add a section to describe that. 
I may change that in the future to be easier to use.

> Is there anything special when providing large ignition files like bootstrap.ign

No. The boostrap.ign works the same ways for masters and workers, our DB doesn't have the limitation I think openstack had/has

Reply:
> For this stage the users will upload the RHCOS qcow image as a template. Later in the survey they will see it and choose it. I add a section to describe that.
I may change that in the future to be easier to use.

So users are allowed to pick a template other than vetted by the installer in IPI??

that is a departure from other platforms..

Reply:
How does the installer does that? guide me here, and I'll asses how much work do I need to make it work like that.

Reply:
on AWS, the installer picks the AMI from the embedded https://github.com/openshift/installer/blob/909f4f8409de9ffe561779bd9f46909a0a14da1d/data/data/rhcos.json#L2-L51 and then creates the machines.

on Azure, the installer picks the vhd URL from embedded https://github.com/openshift/installer/blob/909f4f8409de9ffe561779bd9f46909a0a14da1d/data/data/rhcos.json#L54 -> imports the VHD to percluster storage blob -> Creates an Image using the VHD and creates machines using the Image.

similarly on GCP.

openstack is also in the process to moving to this model, https://github.com/openshift/installer/pull/2473